extensions/wikia/GlobalWatchlist/GlobalWatchlistBot.class.php

```diff
@@ -0,0 +1,432 @@
+		try {
+			$this->checkIfValidUser( $user );
+			$this->checkIfEmailUnSubscribed( $user );
+			$this->checkIfEmailConfirmed( $user );
+			$this->checkIfSubscribedToWeeklyDigest( $user );
+		} catch ( Exception $e ) {
+			if ( $sendLogging ) {
+				WikiaLogger::instance()->info( 'Weekly Digest Skipped', [
+					'reason' => $e->getMessage(),
+					'userID' => $userID
+				]);
+			}
+			return true;
+		}
+		return false;
+	}
+
+	/**
+	 * @param $userID
+	 * @return null|User
+	 */
+	private function getUserObject( $userID ) {
+		global $wgExternalAuthType;
+
+		if ( $wgExternalAuthType ) {
+			$mExtUser = ExternalUser::newFromId( $userID );
+			if ( is_object( $mExtUser ) && ( 0 != $mExtUser->getId() ) ) {
+				$mExtUser->linkToLocal( $mExtUser->getId() );
+				$user = $mExtUser->getLocalUser();
+			} else {
+				$user = null;
+			}
+		} else {
+			$user = User::newFromId ( $userID );
+		}
+
+		return $user;
+	}
+
+	/**
+	 * @param $user
+	 * @throws Exception
+	 */
+	private function checkIfValidUser( $user ) {
+		if ( !$user instanceof User ) {
+			throw new Exception( 'Invalid user object.' );
+		}
+	}
+
+	/**
+	 * @param $user User
+	 * @throws Exception
+	 */
+	private function checkIfEmailUnSubscribed( $user ) {
+		if ( $user->getBoolOption( 'unsubscribed' ) ) {
+			throw new Exception( 'Email is unsubscribed.' );
+		}
+	}
+
+	/**
+	 * @param $user User
+	 * @throws Exception
+	 */
+	private function checkIfEmailConfirmed( $user ) {
+		if ( !$user->isEmailConfirmed() ) {
+			throw new Exception( 'Email is not confirmed.' );
+		}
+	}
+
+	/**
+	 * @param $user User
+	 * @throws Exception
+	 */
+	private function checkIfSubscribedToWeeklyDigest( $user ) {
+		if ( !$user->getBoolOption( 'watchlistdigest' ) ) {
+			throw new Exception( 'Not subscribed to weekly digest' );
+		}
+	}
+
+	/**
+	 * @param $userID
+	 */
+	private function clearUserFromGlobalWatchlist( $userID ) {
+		$task = new GlobalWatchlistTask();
+		( new AsyncTaskList() )
+			->wikiId( F::app()->wg->CityId )
+			->add( $task->call( 'clearGlobalWatchlistAll', $userID ) )
+			->queue();
+	}
+
+	/**
+	 * send email
+	 */
+	private function sendMail( $iUserId, $aDigestData, $isDigestLimited ) {
+		$oUser = User::newFromId( $iUserId );
+		$oUser->load();
+
+		$sEmailSubject = $this->getLocalizedMsg( 'globalwatchlist-digest-email-subject', $oUser->getOption( 'language' ) );
+		list( $sEmailBody, $sEmailBodyHTML ) = $this->composeMail( $oUser, $aDigestData, $isDigestLimited );
+
+		$sFrom = 'Wikia <community@wikia.com>';
+		// yes this needs to be a MA object, not string (the docs for sendMail are wrong)
+		$oReply = new MailAddress( 'noreply@wikia.com' );
+
+		$oUser->sendMail( $sEmailSubject, $sEmailBody, $sFrom, $oReply, 'GlobalWatchlist', $sEmailBodyHTML );
+
+		WikiaLogger::instance()->info( 'Weekly Digest Sent', [ 'userID' => $iUserId ] );
+	}
+
+	/**
+	 * compose digest email for user
+	 */
+	function composeMail ( $oUser, $aDigestsData, $isDigestLimited ) {
```

Comment:
I know this is moved, but this is also a giant function. Can it be broken up easily?